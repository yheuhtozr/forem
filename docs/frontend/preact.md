---
title: Preact
---

# Preact

[Preact](https://preactjs.com/) is an alternative to React with the same modern
API.

Preact components are packaged using [Webpacker](/frontend/webpacker) and the
Preact code is located in `app/javascript`.

Preact components get loaded via webpacker's helper function
`javascript_packs_with_chunks_tag`.

## PropTypes

Preact supports
[PropTypes](https://reactjs.org/docs/typechecking-with-proptypes.html). When
creating Preact components, please ensure that you have defined your PropTypes.

### Common PropTypes

Using PropTypes can be repetitive. Some duplication is normal, like when a
PropType is a string or a number. But for commonly-used PropTypes, like the user
entity, you can use the provided common PropTypes, located in the
`/app/javascript/common-prop-types`, as shown below.

```javascript
import PropTypes from 'prop-types';

export const userPropTypes = PropTypes.shape({
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  profile_image_url: PropTypes.string.isRequired,
  summary: PropTypes.string.isRequired,
});
```

#### Using Common PropTypes

Common PropTypes are imported just like any other
[JavaScript Module](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules).
For example, here are two scenarios where a component needs to use the
`tagPropTypes`.

In the example below, our component `SomeComponentUsingTags` has a `tags` prop,
which is an array of the tag entity. PropTypes have a built-in method called
`arrayOf` that allows you to define a prop as an array of something. In our
case, this is the tag entity, so we can use the `tagPropTypes` PropType.

```jsx
import { h } from 'preact';
import PropTypes from 'prop-types';
import { tagPropTypes } from '../../../components/common-prop-types';

const SomeComponentUsingTags = ({ tags = [] }) => (
  <ul>
    {tags.map((tag) => (
      <li key={tag.id}>{tag.name}</li>
    ))}
  </ul>
);

SomeComponentUsingTags.displayName = 'SomeComponentUsingTags';
SomeComponentUsingTags.propTypes = {
  tags: PropTypes.arrayOf(tagPropTypes).isRequired,
};
```

In the following example, the `SomeComponentUsingOneTag` component has a `tag`
prop representing a single tag. In this case, we can just the `tagPropTypes` on
their own to represent the shape of the `tag` prop.

```jsx
import { h } from 'preact';
import { tagPropTypes } from '../../../components/common-prop-types';

const SomeComponentUsingOneTag = ({ tag }) => <li key={tag.id}>{tag.name}</li>;

SomeComponentUsingOneTag.displayName = 'SomeComponentUsingTags';
SomeComponentUsingOneTag.propTypes = {
  tag: tagPropTypes.isRequired,
};
```
