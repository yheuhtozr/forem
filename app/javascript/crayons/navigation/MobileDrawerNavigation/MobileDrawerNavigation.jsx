import { h, Fragment } from 'preact';
import { useState } from 'preact/hooks';
import PropTypes from 'prop-types';
import { MobileDrawer } from '@crayons/MobileDrawer';
import { Button } from '@crayons/Button';

const OverflowIcon = () => (
  <svg width="24" height="24" xmlns="http://www.w3.org/2000/svg">
    <path
      fill-rule="evenodd"
      clip-rule="evenodd"
      d="M7 12a2 2 0 11-4 0 2 2 0 014 0zm7 0a2 2 0 11-4 0 2 2 0 014 0zm5 2a2 2 0 100-4 2 2 0 000 4z"
    />
  </svg>
);

const CheckIcon = () => (
  <svg
    aria-hidden="true"
    className="check-icon"
    fill="currentColor"
    width="24"
    height="24"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path d="M10 15.172l9.192-9.193 1.415 1.414L10 18l-6.364-6.364 1.414-1.414 4.95 4.95z" />
  </svg>
);

/**
 * Renders a page heading with a button which activates a <MobileDrawer /> with a list of the given navigation links.
 *
 * @param {object} props
 * @param {number} headingLevel The level of heading to render as the page title (e.g. 1-6)
 * @param {string} navigationTitle The title to be used for the navigation element (e.g. 'Feed timeframes')
 * @param {Array} navigationLinks An array of navigationLink objects to display
 *
 * @example
 * <MobileDrawerNavigation
 *   headingLevel={2}
 *   navigationTitle="Feed options"
 *   navigationLinks={links} />
 */
export const MobileDrawerNavigation = ({
  headingLevel,
  navigationTitle,
  navigationLinks,
}) => {
  const [isDrawerOpen, setIsDrawerOpen] = useState(false);
  const currentPage = navigationLinks.find((item) => item.isCurrentPage);

  const Heading = `h${headingLevel}`;

  return (
    <Fragment>
      <div className="flex justify-between">
        <Heading>{currentPage.displayName}</Heading>
        <Button
          aria-label={navigationTitle}
          icon={OverflowIcon}
          size="s"
          contentType="icon"
          variant="ghost"
          onClick={() => setIsDrawerOpen(true)}
        />
      </div>

      {isDrawerOpen && (
        <MobileDrawer
          title={navigationTitle}
          onClose={() => setIsDrawerOpen(false)}
        >
          <nav aria-label={navigationTitle} className="drawer-navigation">
            <ul className="list-none">
              {navigationLinks.map((linkDetails) => {
                const { url, isCurrentPage, displayName } = linkDetails;
                return (
                  <li
                    key={`link-${url}`}
                    className="drawer-navigation__item py-2"
                  >
                    <a href={url} aria-current={isCurrentPage ? 'page' : null}>
                      {displayName}
                    </a>
                    <CheckIcon />
                  </li>
                );
              })}
            </ul>
          </nav>
          <Button
            variant="secondary"
            className="w-100 mt-4"
            onClick={() => setIsDrawerOpen(false)}
          >
            Cancel
          </Button>
        </MobileDrawer>
      )}
    </Fragment>
  );
};

MobileDrawerNavigation.propTypes = {
  headingLevel: PropTypes.oneOf([1, 2, 3, 4, 5, 6]).isRequired,
  navigationTitle: PropTypes.string.isRequired,
  navigationLinks: PropTypes.arrayOf(
    PropTypes.shape({
      url: PropTypes.string,
      isCurrentPage: PropTypes.bool,
      displayName: PropTypes.string,
    }),
  ).isRequired,
};
