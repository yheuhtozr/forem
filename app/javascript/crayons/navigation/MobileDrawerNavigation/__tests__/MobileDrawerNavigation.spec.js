import { h } from 'preact';
import { axe } from 'jest-axe';
import '@testing-library/jest-dom';
import { render, waitFor } from '@testing-library/preact';
import { MobileDrawerNavigation } from '../MobileDrawerNavigation';

describe('<MobileDrawerNavigation />', () => {
  const testLinks = [
    { url: '/#1', displayName: 'Link 1', isCurrentPage: true },
    { url: '/#2', displayName: 'Link 2', isCurrentPage: false },
    { url: '/#3', displayName: 'Link 3', isCurrentPage: false },
  ];

  it('should have no a11y violations when closed', async () => {
    const { container } = render(
      <MobileDrawerNavigation
        navigationTitle="Test navigation"
        headingLevel={1}
        navigationLinks={testLinks}
      />,
    );

    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should have no a11y violations when open', async () => {
    const { container, getByRole } = render(
      <MobileDrawerNavigation
        navigationTitle="Test navigation"
        headingLevel={1}
        navigationLinks={testLinks}
      />,
    );

    getByRole('button', { name: 'Test navigation' }).click();
    await waitFor(() => getByRole('navigation', { name: 'Test navigation' }));

    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should render a heading with a button when closed', () => {
    const { container } = render(
      <MobileDrawerNavigation
        navigationTitle="Test navigation"
        headingLevel={1}
        navigationLinks={testLinks}
      />,
    );

    expect(container.innerHTML).toMatchSnapshot();
  });

  it('should render a navigation with a checkmark for current page when open', async () => {
    const { container, getByRole } = render(
      <MobileDrawerNavigation
        navigationTitle="Test navigation"
        headingLevel={1}
        navigationLinks={testLinks}
      />,
    );

    getByRole('button', { name: 'Test navigation' }).click();
    await waitFor(() => getByRole('navigation', { name: 'Test navigation' }));
    expect(container.innerHTML).toMatchSnapshot();
  });

  it('should render all links', async () => {
    const { getByRole } = render(
      <MobileDrawerNavigation
        navigationTitle="Test navigation"
        headingLevel={1}
        navigationLinks={testLinks}
      />,
    );

    getByRole('button', { name: 'Test navigation' }).click();
    await waitFor(() => getByRole('navigation', { name: 'Test navigation' }));

    expect(getByRole('link', { name: 'Link 1' })).toBeInTheDocument();
    expect(getByRole('link', { name: 'Link 2' })).toBeInTheDocument();
    expect(getByRole('link', { name: 'Link 3' })).toBeInTheDocument();
  });
});
