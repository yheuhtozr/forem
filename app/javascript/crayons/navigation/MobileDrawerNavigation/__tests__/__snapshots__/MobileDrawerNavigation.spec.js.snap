// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MobileDrawerNavigation /> should render a heading with a button when closed 1`] = `"<div class=\\"flex justify-between\\"><h1>Link 1</h1><button class=\\"crayons-btn crayons-btn--ghost crayons-btn--s crayons-btn--icon\\" type=\\"button\\" aria-label=\\"Test navigation\\"><svg width=\\"24\\" height=\\"24\\" xmlns=\\"http://www.w3.org/2000/svg\\"><path fill-rule=\\"evenodd\\" clip-rule=\\"evenodd\\" d=\\"M7 12a2 2 0 11-4 0 2 2 0 014 0zm7 0a2 2 0 11-4 0 2 2 0 014 0zm5 2a2 2 0 100-4 2 2 0 000 4z\\"></path></svg></button></div>"`;

exports[`<MobileDrawerNavigation /> should render a navigation with a checkmark for current page when open 1`] = `"<div class=\\"flex justify-between\\"><h1>Link 1</h1><button class=\\"crayons-btn crayons-btn--ghost crayons-btn--s crayons-btn--icon\\" type=\\"button\\" aria-label=\\"Test navigation\\"><svg width=\\"24\\" height=\\"24\\" xmlns=\\"http://www.w3.org/2000/svg\\"><path fill-rule=\\"evenodd\\" clip-rule=\\"evenodd\\" d=\\"M7 12a2 2 0 11-4 0 2 2 0 014 0zm7 0a2 2 0 11-4 0 2 2 0 014 0zm5 2a2 2 0 100-4 2 2 0 000 4z\\"></path></svg></button></div><div class=\\"crayons-mobile-drawer\\"><div class=\\"crayons-mobile-drawer__overlay\\"></div><div class=\\"crayons-mobile-drawer__content\\" role=\\"dialog\\" aria-modal=\\"true\\" aria-label=\\"Test navigation\\"><nav aria-label=\\"Test navigation\\" class=\\"drawer-navigation\\"><ul class=\\"list-none\\"><li class=\\"drawer-navigation__item py-2\\"><a href=\\"/#1\\" aria-current=\\"page\\">Link 1</a><svg aria-hidden=\\"true\\" class=\\"check-icon\\" fill=\\"currentColor\\" width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" xmlns=\\"http://www.w3.org/2000/svg\\"><path d=\\"M10 15.172l9.192-9.193 1.415 1.414L10 18l-6.364-6.364 1.414-1.414 4.95 4.95z\\"></path></svg></li><li class=\\"drawer-navigation__item py-2\\"><a href=\\"/#2\\">Link 2</a><svg aria-hidden=\\"true\\" class=\\"check-icon\\" fill=\\"currentColor\\" width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" xmlns=\\"http://www.w3.org/2000/svg\\"><path d=\\"M10 15.172l9.192-9.193 1.415 1.414L10 18l-6.364-6.364 1.414-1.414 4.95 4.95z\\"></path></svg></li><li class=\\"drawer-navigation__item py-2\\"><a href=\\"/#3\\">Link 3</a><svg aria-hidden=\\"true\\" class=\\"check-icon\\" fill=\\"currentColor\\" width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" xmlns=\\"http://www.w3.org/2000/svg\\"><path d=\\"M10 15.172l9.192-9.193 1.415 1.414L10 18l-6.364-6.364 1.414-1.414 4.95 4.95z\\"></path></svg></li></ul></nav><button class=\\"crayons-btn crayons-btn--secondary w-100 mt-4\\" type=\\"button\\">Cancel</button></div></div>"`;
